### PR TITLE
Proof blocks

### PR DESCRIPTION
**Pull request by SethPoulsen**
Added the DAG based grader for grading proofs into `pl-order-blocks`. 

Almost ready to go, but will be waiting to merge until https://github.com/PrairieLearn/PrairieLearn/pull/3837 is done.

### CODE REVIEW

**Comment by SethPoulsen** [comment]
- - -
   ***I have reviewed and hereby sign the CLA***
   - - -

**Comment by SethPoulsen** [comment]
@nwalters512 how can I make this CLA bot happy? Did I copy-paste the thing wrong or something? Thanks.

**Comment by nwalters512** [comment]
@SethPoulsen sorry it wasn't clear - you need to post without the formatting, just the text string!

**Comment by SethPoulsen** [comment]
I have reviewed and hereby sign the CLA

**Comment by SethPoulsen** [comment]
recheck

**Comment by SethPoulsen** [comment]
I've just noticed that with this element (including the original grading modes currently in use, before this PR) will show the partial credit that you *would have* gotten on a question, even if the question has `"partialCredit": false`. Is this behavior consistent with other PL elements or should we change it? 

From my end, I don't particularly like it because (1) I think its confusing in general and (2) in the case of this element, it gives the user a hint about how many total blocks there are supposed to be. 

image

File: elements/pl-order-blocks/pl-order-blocks.py [hunk 1]
```diff
@@ -53,38 +48,67 @@ def prepare(element_html, data):
...
-        index += 1
+
+    index = 0
+    group_counter = 0
+    for html_tags in element:  # iterate through the html tags inside pl-order-blocks
+        if html_tags.tag is lxml.etree.Comment:
+            continue
+        elif html_tags.tag == 'pl-block-group':
+            if grading_method != 'dag':
+                raise Exception('Block groups only supported in the DAG grading mode.')
```
Comment (mwest1066) [inline elements/pl-order-blocks/pl-order-blocks.py:101]:
```suggestion
                raise Exception('Block groups only supported in the "dag" grading mode.')
```

File: elements/pl-order-blocks/pl-order-blocks.py [hunk 2]
```diff
@@ -138,6 +162,7 @@ def render(element_html, data):
 
         mcq_options = data['params'][answer_name]
         mcq_options = filter_multiple_from_array(mcq_options, ['inner_html', 'uuid'])
+        # mcq_options = [opt.strip() for opt in mcq_options] Still needed?
```
Comment (mwest1066) [inline elements/pl-order-blocks/pl-order-blocks.py:165]: Either delete this comment or fix it?

File: elements/pl-order-blocks/pl-order-blocks.py [hunk 3]
```diff
@@ -343,6 +373,25 @@ def grade(element_html, data):
...
+        correctness, first_wrong = grade_dag(order, depends_graph, group_belonging)
+
+        if correctness == len(depends_graph.keys()):
+            final_score = 1
+        elif correctness < len(depends_graph.keys()):
+            final_score = float(correctness) / len(depends_graph.keys())
+            if first_wrong == -1:
+                feedback = 'Your answer is correct so far, but it is incomplete.'
+            else:
+                feedback = r"""Your answer is incorrect starting at <span style="color:red;">line number """ + str(first_wrong + 1) + \
```
Comment (mwest1066) [inline elements/pl-order-blocks/pl-order-blocks.py:390]: `line` or `block`?

Comment (SethPoulsen) [reply elements/pl-order-blocks/pl-order-blocks.py]: changed to block

File: elements/pl-order-blocks/pl-order-blocks.py (hunk 2 above)
Comment (SethPoulsen) [reply elements/pl-order-blocks/pl-order-blocks.py]: deleted

File: elements/pl-order-blocks/pl-order-blocks.py (hunk 1 above)
Comment (SethPoulsen) [reply elements/pl-order-blocks/pl-order-blocks.py]: fixed

**Comment by SethPoulsen** [comment]
@mwest1066 I'm ready for you to look over this again, the only thing I'm not positive about is if I want the feedback text on by default or not.

**Review by mwest1066 — APPROVED** [review]

**Comment by mwest1066** [comment]
> @mwest1066 I'm ready for you to look over this again, the only thing I'm not positive about is if I want the feedback text on by default or not.

LGTM. Ping me again once you've decided on default feedback behavior.

**Comment by SethPoulsen** [comment]
I've chosen to leave no feedback by default for consistency with other grading modes. `feedback: 'first-wrong'` will give the feedback if you are using `grading-mode:'dag'`.

Any other feedback option will cause an error. 

All good from my side, please merge when ready.

**Review by mwest1066 — APPROVED** [review]

**Pull request merged**